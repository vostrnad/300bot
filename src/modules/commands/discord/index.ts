import alts from './alts'
import certs from './certs'
import cooldown from './cooldown'
import hangman from './hangman'
import outfitstats from './outfitstats'
import playerstats from './playerstats'
import population from './population'
import prefix from './prefix'
import russianroulette from './russianroulette'
import track from './track'
import urbandictionary from './urbandictionary'

export const discordCommands = [
  alts,
  certs,
  cooldown,
  hangman,
  playerstats,
  population,
  prefix,
  russianroulette,
  track,
  urbandictionary,
  outfitstats,
]
