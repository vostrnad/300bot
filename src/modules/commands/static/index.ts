import { StaticCommand } from './StaticCommand'

export const staticCommands: StaticCommand[] = [
  new StaticCommand({
    keyword: 'rickroll',
    description: 'get rickrolled',
    help: 'Usage: `{prefix}rickroll` - get rickrolled',
    response: 'Never gonna give you up, never gonna let you down.',
  }),
  new StaticCommand({
    keyword: 'hi',
    description: 'says Hi!',
    help: 'Usage: `{prefix}hi` - says Hi!',
    response: 'Hi!',
  }),
]
