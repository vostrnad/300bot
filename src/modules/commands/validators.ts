import { ArgumentNumberError, ArgumentRangeError } from '@app/errors'

export const validateArgumentNumber = (
  actual: number,
  required: number,
): void => {
  if (actual !== required) {
    throw new ArgumentNumberError(actual, required)
  }
}

export const validateArgumentRange = (
  actual: number,
  min?: number,
  max?: number,
): void => {
  if (
    (min !== undefined && actual < min) ||
    (max !== undefined && actual > max)
  ) {
    throw new ArgumentRangeError(actual, min, max)
  }
}
