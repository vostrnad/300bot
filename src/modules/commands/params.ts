import discord from 'discord.js'

export type Settings = {
  prefix: string
  outfitId?: string
}

export type SettingsParams = {
  settings: Settings
  updateSettings: <K extends keyof Settings>(
    key: K,
    value: Settings[K],
  ) => void | Promise<void>
}

export type DiscordParams = SettingsParams & {
  message: discord.Message
}
