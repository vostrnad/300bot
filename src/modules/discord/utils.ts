import discord from 'discord.js'
import { log } from '@app/utils/log'

export const getTextChannel = (
  client: discord.Client,
  id: string,
): discord.TextChannel | null => {
  const channel = client.channels.resolve(id)
  if (channel instanceof discord.TextChannel) {
    return channel
  }
  return null
}

export const getDMChannel = async (
  client: discord.Client,
  id: string,
): Promise<discord.DMChannel | null> => {
  const channel = await client.channels.fetch(id)
  if (channel instanceof discord.DMChannel) {
    return channel
  }
  return null
}

export const formatWithEmojis = (
  channel: discord.Channel,
  message: string,
): string => {
  // eslint-disable-next-line unicorn/no-unsafe-regex
  const emojiExpressionRegex = /{emoji:([^|]*)(?:\|([^}]*))?}/g

  if (channel instanceof discord.TextChannel) {
    return message.replace(
      emojiExpressionRegex,
      (_match, emojiName, fallbackText) => {
        const emoji = channel.guild.emojis.cache.find(
          ({ name }) => name === emojiName,
        )
        if (emoji) {
          return emoji.toString()
        } else if (typeof fallbackText === 'string') {
          return fallbackText
        } else {
          return ''
        }
      },
    )
  } else {
    return message.replace(
      emojiExpressionRegex,
      (_match, _emojiName, fallbackText) => {
        if (typeof fallbackText === 'string') {
          return fallbackText
        } else {
          return ''
        }
      },
    )
  }
}

export const removeReaction = async (
  reaction: discord.MessageReaction,
  user: discord.User,
): Promise<void> => {
  try {
    await reaction.users.remove(user)
  } catch (error) {
    log.error('Error removing reaction:', error)
  }
}
