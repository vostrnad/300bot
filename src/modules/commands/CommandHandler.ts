import fuzzySet from 'fuzzyset.js'
import { CustomError } from '@app/errors'
import { log } from '@app/utils/log'

export interface CommandMessage<T = unknown> {
  text: string
  author: MessageAuthor
  reply: ReplyToMessage
  raw: T
}

type ReplyToMessage = (text: string) => void

interface MessageAuthor {
  /** Unique user identifier. */
  id: string
  /** Favored representation of user's name. */
  displayName: string
  /** String that can be used to mention user. */
  mention: string
  /** Determines if user has admin/moderator permissions. */
  admin: boolean
}

export interface CommandCallbackData<T = unknown> {
  alias: string
  args: string[]
  reply: ReplyToMessage
  author: MessageAuthor
  env: {
    handler: CommandHandler<never>
    command: Command<never>
  }
  raw: T
}

export type CommandCallback<T = unknown> = (
  data: CommandCallbackData<T>,
) => Promise<void> | void

interface CommandOptions {
  /**
   * If defined, Limits the number of arguments and joins all excessive
   * text into the last argument.
   */
  lastArgNumber?: number
  /**
   * If true, the command must be prefixed. If false, the command will
   * also work without prefix.
   */
  requirePrefix: boolean
  /**
   * If true, the first argument has to be separated by a space. If false,
   * the first argument does not have to be separated by a space.
   */
  firstArgSpace: boolean
  /**
   * If true, the command will not show in help.
   */
  hidden: boolean
}

export interface CommandConfig<T = unknown> {
  readonly keyword: string
  readonly alias?: string[]
  /** Short description of the command in lowercase. */
  readonly description: string
  /** Usage info on the command. `{prefix}` is replaced with the prefix. */
  readonly help: string
  readonly callback: CommandCallback<T>
  readonly options?: Partial<CommandOptions>
}

export class Command<T = unknown> {
  private static readonly DEFAULT_OPTIONS: CommandOptions = {
    lastArgNumber: undefined,
    requirePrefix: true,
    firstArgSpace: true,
    hidden: false,
  }

  public readonly keyword: string
  public readonly alias: string[]
  public readonly description: string
  public readonly callback: CommandCallback<T>
  public readonly options: CommandOptions

  private readonly _help: string

  constructor(config: CommandConfig<T>) {
    const { keyword, alias, description, help, callback, options } = config
    this.keyword = keyword
    this.alias = alias || []
    this.description = description
    this.callback = callback
    this.options = {
      ...Command.DEFAULT_OPTIONS,
      ...options,
    }
    this._help = help
  }

  public getHelp(handler: CommandHandler<T>): string {
    return this._help.replace(/{prefix}/g, handler.prefix)
  }
}

export interface CommandHandlerConfig<T = unknown> {
  commands: Array<Command<T>>
  prefix: string
  /**
   * Allows brackets in arguments, which is what users
   * sometimes do when misunderstanding the command help.
   */
  allowArgumentBrackets?: boolean
}

export class CommandHandler<T = unknown> {
  public readonly prefix: string
  private readonly _commands: Array<Command<T>>
  private readonly _allowArgumentBrackets: boolean

  constructor(config: CommandHandlerConfig<T>) {
    this.prefix = config.prefix
    this._commands = config.commands
    this._allowArgumentBrackets = config.allowArgumentBrackets || false
  }

  public getPublicCommands(): Array<Command<T>> {
    return this._commands.filter((command) => !command.options.hidden)
  }

  public getCommandHelp(keyword: string): string | null {
    const command = this.getCommand(keyword)
    if (command !== null) {
      return command.getHelp(this)
    } else {
      return null
    }
  }

  /**
   * Processes a text message. Returns true if a command was detected
   * and acted upon, false otherwise.
   */
  public async process(message: CommandMessage<T>): Promise<boolean> {
    const prefix = this.prefix
    for (const command of this._commands) {
      const commandAliases: string[] = [command.keyword, ...command.alias]

      for (const alias of commandAliases) {
        /** Command string without prefix. */
        let commandString: string

        if (message.text.startsWith(`${prefix}${alias}`)) {
          commandString = message.text.slice(prefix.length)
        } else if (
          command.options.requirePrefix === false &&
          message.text.startsWith(alias)
        ) {
          commandString = message.text
        } else {
          continue
        }

        let argString: string

        if (commandString === alias) {
          argString = ''
        } else if (commandString.startsWith(`${alias} `)) {
          argString = commandString.slice(alias.length + 1)
        } else if (command.options.firstArgSpace === false) {
          argString = commandString.slice(alias.length)
        } else {
          continue
        }

        const rawArgs = argString.split(' ')
        let filteredArgs: string[] = []

        if (!command.options.lastArgNumber) {
          // filter arguments from spaces
          filteredArgs = rawArgs.filter((value: string) => {
            return value.length > 0
          })
        } else {
          // fill arguments with max args and then join the rest
          for (let i = 0; i < rawArgs.length; i++) {
            const arg: string = rawArgs[i]
            if (filteredArgs.length < command.options.lastArgNumber - 1) {
              if (arg.length > 0) {
                filteredArgs.push(arg)
              }
            } else {
              const lastArg = rawArgs.slice(i).join(' ').trim()
              if (lastArg.length > 0) {
                filteredArgs.push(lastArg)
              }
              break
            }
          }
        }

        if (this._allowArgumentBrackets) {
          // remove < and > from arguments
          filteredArgs = filteredArgs.map((arg: string) => {
            if (arg.length > 2 && arg.startsWith('<') && arg.endsWith('>')) {
              return arg.slice(1, -1)
            } else {
              return arg
            }
          })
        }

        try {
          await command.callback({
            alias,
            args: filteredArgs,
            reply: message.reply,
            author: message.author,
            env: {
              handler: this,
              command,
            },
            raw: message.raw,
          })
        } catch (e) {
          if (e instanceof CustomError) {
            message.reply(`Error: ${e.message}.`)
          } else {
            message.reply('An error occurred while processing the command.')
            log.error('Command error:', e)
          }
        }

        return true
      }
    }

    if (message.text.startsWith(`${prefix}`)) {
      const commandsAliases: string[] = []
      this._commands.forEach((command) => {
        commandsAliases.push(command.keyword, ...command.alias)
      })

      const fz = fuzzySet(commandsAliases)
      const match = fz.get(message.text.slice(prefix.length))
      if (!match) {
        // if match is less than 33% null is returned by default in the module
        return false
      }

      const distthreshold = 0.6 // 60% of the keyword/alias has to be right for a suggestion to be relevant
      if (match[0][0] < 1 - distthreshold) {
        message.reply(
          `Unknown command, use **${prefix}help** to get a list of all commands.`,
        )
      } else {
        message.reply(
          `Unknown command, did you mean **${prefix}${match[0][1]}**?`,
        )
      }
    }
    return false
  }

  private getCommand(keyword: string) {
    const found = this._commands.find((command) => {
      return command.keyword === keyword || command.alias.includes(keyword)
    })
    if (found !== undefined) {
      return found
    } else {
      return null
    }
  }
}
