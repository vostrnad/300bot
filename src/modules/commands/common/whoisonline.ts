import { OutfitAliasNotFoundError } from '@app/errors'
import { sentence } from '@app/utils/language'
import { Command } from '@commands/CommandHandler'
import { SettingsParams } from '@commands/params'
import {
  validateArgumentRange,
  validateDefaultOutfitId,
} from '@commands/validators'
import { censusApi } from '@planetside/CensusApi'
import { validateOutfitAlias } from '@planetside/validators'

export default new Command<SettingsParams>({
  keyword: 'whoisonline',
  description: 'check who is online',
  help: 'Usage:\n`{prefix}whoisonline` - checks who is online\n`{prefix}whoisonline <outfit tag>` - list online members of a specific outfit',
  callback: async ({ args, reply, env }) => {
    validateArgumentRange(args.length, 0, 1)

    if (args.length === 0) {
      validateDefaultOutfitId(env.settings.outfitId, env.handler.prefix)

      const names = await censusApi.getOnlineOutfitMembers(
        env.settings.outfitId,
      )

      if (names.length === 0) {
        reply('No outfit members are online.')
      } else if (names.length === 1) {
        reply(`**${names[0]}** is the only online outfit member.`)
      } else {
        reply(
          `These ${names.length} outfit members are online: ${sentence(
            names.map((name) => `**${name}**`),
          )}.`,
        )
      }
    } else {
      const alias = args[0]
      validateOutfitAlias(alias)
      const outfit = await censusApi.getOutfit({
        aliasLower: alias.toLowerCase(),
      })
      if (!outfit) throw new OutfitAliasNotFoundError()
      const names = await censusApi.getOnlineOutfitMembers(outfit.outfitId)

      if (names.length === 0) {
        reply(`No members of **${outfit.name}** are online.`)
      } else if (names.length === 1) {
        reply(
          `**${names[0]}** is the only online member of **${outfit.name}**.`,
        )
      } else {
        reply(
          `These ${names.length} members of **${
            outfit.name
          }** are online: ${sentence(names.map((name) => `**${name}**`))}.`,
        )
      }
    }
  },
})
