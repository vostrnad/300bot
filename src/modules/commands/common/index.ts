import { Command } from '@commands/CommandHandler'
import brudistance from './brudistance'
import certs from './certs'
import coinflip from './coinflip'
import diceroll from './diceroll'
import directive from './directive'
import hello from './hello'
import help from './help'
import isbruonline from './isbruonline'
import math from './math'
import membercount from './membercount'
import outfitstats from './outfitstats'
import playersearch from './playersearch'
import playerstats from './playerstats'
import population from './population'
import recentstats from './recentstats'
import restartsocket from './restartsocket'
import uptime from './uptime'
import whoisleading from './whoisleading'
import whoisonline from './whoisonline'

export const commonCommands: Command[] = [
  brudistance,
  certs,
  coinflip,
  diceroll,
  directive,
  hello,
  help,
  isbruonline,
  math,
  membercount,
  outfitstats,
  playersearch,
  playerstats,
  population,
  recentstats,
  restartsocket,
  uptime,
  whoisleading,
  whoisonline,
]
