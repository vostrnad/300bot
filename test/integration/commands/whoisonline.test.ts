import whoisonline from '@commands/common/whoisonline'
import { getCommandRunner } from '@test/utils/commands'
import { generate } from '@test/utils/data'
import { mockCensusApi } from '@test/utils/planetside'

describe('whoisonline', () => {
  const runCommand = getCommandRunner(whoisonline)

  it('should work when API response is empty', async () => {
    const requestData = mockCensusApi({ outfit_member_list: [] })
    const reply = await runCommand('+whoisonline')
    expect(requestData).toMatchSnapshot()
    expect(reply).toEqual('No outfit members are online.')
  })

  it('should work when everyone is offline', async () => {
    const list = generate(3, (index) => ({
      character_id: `id${index}`,
      character: {
        name: { first: `TestMember${index}` },
        online_status: '0',
      },
    }))
    const requestData = mockCensusApi({ outfit_member_list: list })
    const reply = await runCommand('+whoisonline')
    expect(requestData).toMatchSnapshot()
    expect(reply).toEqual('No outfit members are online.')
  })

  it('should work when one member is online', async () => {
    const list = generate(3, (index) => ({
      character_id: `id${index}`,
      character: {
        name: { first: `TestMember${index}` },
        online_status: index === 1 ? '1' : '0',
      },
    }))
    const requestData = mockCensusApi({ outfit_member_list: list })
    const reply = await runCommand('+whoisonline')
    expect(requestData).toMatchSnapshot()
    expect(reply).toEqual('**TestMember1** is the only online outfit member.')
  })

  it('should work when some members are online', async () => {
    const list = generate(6, (index) => ({
      character_id: `id${index}`,
      character: {
        name: { first: `TestMember${index}` },
        online_status: index % 2 === 0 ? '0' : '1',
      },
    }))
    const requestData = mockCensusApi({ outfit_member_list: list })
    const reply = await runCommand('+whoisonline')
    expect(requestData).toMatchSnapshot()
    expect(reply).toEqual(
      'These 3 outfit members are online: **TestMember1**, **TestMember3** and **TestMember5**.',
    )
  })
})
