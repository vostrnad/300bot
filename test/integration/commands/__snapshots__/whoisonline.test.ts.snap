// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`whoisonline should work when API response is empty 1`] = `
Object {
  "pathname": "/s:test/get/ps2:v2/outfit_member",
  "searchParams": Object {
    "c:join": Array [
      "character_name^on:character_id^inject_at:character^show:name.first",
      "characters_online_status^on:character_id^inject_at:character^show:name.first",
    ],
    "c:limit": "65535",
    "c:show": "character_id",
    "outfit_id": "37581022577821398",
  },
}
`;

exports[`whoisonline should work when everyone is offline 1`] = `
Object {
  "pathname": "/s:test/get/ps2:v2/outfit_member",
  "searchParams": Object {
    "c:join": Array [
      "character_name^on:character_id^inject_at:character^show:name.first",
      "characters_online_status^on:character_id^inject_at:character^show:name.first",
    ],
    "c:limit": "65535",
    "c:show": "character_id",
    "outfit_id": "37581022577821398",
  },
}
`;

exports[`whoisonline should work when one member is online 1`] = `
Object {
  "pathname": "/s:test/get/ps2:v2/outfit_member",
  "searchParams": Object {
    "c:join": Array [
      "character_name^on:character_id^inject_at:character^show:name.first",
      "characters_online_status^on:character_id^inject_at:character^show:name.first",
    ],
    "c:limit": "65535",
    "c:show": "character_id",
    "outfit_id": "37581022577821398",
  },
}
`;

exports[`whoisonline should work when some members are online 1`] = `
Object {
  "pathname": "/s:test/get/ps2:v2/outfit_member",
  "searchParams": Object {
    "c:join": Array [
      "character_name^on:character_id^inject_at:character^show:name.first",
      "characters_online_status^on:character_id^inject_at:character^show:name.first",
    ],
    "c:limit": "65535",
    "c:show": "character_id",
    "outfit_id": "37581022577821398",
  },
}
`;
